File: bin/benchmarks/projects/freehep/base/freehep-xml/src/main/java/org/freehep/wbxml/DefaultExtensionHandler.java
// Copyright FreeHEP, 2007
package org.freehep.wbxml;

import java.io.DataInputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import org.xml.sax.SAXException;

public class DefaultExtensionHandler implements ExtensionHandler {

	private List strings = new ArrayList();

	public void ext(int i, int tagID, int attributeID,
			MutableAttributes atts, List value) throws SAXException {
		throw new SAXException("Unknown extension: Ext " + i + " tagID: " + tagID
				+ " attributeID: " + attributeID);
	}

	public void extI(int i, String s, int tagID,
			int attributeID, MutableAttributes atts, List value) throws SAXException {
		if (i==0) {
			atts.set(attributeID, s);
			strings.add(s);
		} else {
			throw new SAXException("Unknown extension: ExtI " + i + " " + s + " tagID: "
				+ tagID + " attributeID: " + attributeID);
		}
	}

	public void extT(int i, int index, int tagID, int attributeID,
			MutableAttributes atts, List value) throws SAXException {
		if (i==0) {
			atts.set(attributeID, (String)strings.get(index));
		} else {
			throw new SAXException("Unknown extension: ExtT " + i + " " + index + " tagID: "
				+ tagID + " attributeID: " + attributeID);
		}
	}

	public void opaque(int len, DataInputStream in, int tagID,
			int attributeID, MutableAttributes atts, List value)
			throws IOException, SAXException {
		int type = in.readByte();
		len--;
		switch(type) {
		case Attributes.BOOLEAN:
			if (len == 1) {
				atts.set(attributeID, in.readByte() != 0);
				len--;
			} else {
				boolean[] array = new boolean[len];
				for (int i=0; i<array.length; i++) {
					array[i] = in.readByte() != 0;
					len--;
				}
				atts.set(attributeID, array);
			}
			break;
		case Attributes.BYTE:
			if (len == 1) {
				atts.set(attributeID, in.readByte());
				len--;
			} else {
				byte[] array = new byte[len];
				for (int i=0; i<array.length; i++) {
					array[i] = in.readByte();
					len--;
				}
				atts.set(attributeID, array);
			}
			break;
		case Attributes.CHAR:
			if (len == 2) {
				atts.set(attributeID, in.readChar());
				len-=2;
			} else {
				char[] array = new char[len/2];
				for (int i=0; i<array.length; i++) {
					array[i] = in.readChar();
					len-=2;
				}
				atts.set(attributeID, array);
			}
			break;
		case Attributes.DOUBLE:
			if (len == 8) {
				atts.set(attributeID, in.readDouble());
				len -= 8;
			} else {
				double[] array = new double[len/8];
				for (int i=0; i<array.length; i++) {
					array[i] = in.readDouble();
					len-=8;
				}
				atts.set(attributeID, array);
			}
			break;
		case Attributes.FLOAT:
			if (len == 4) {
				atts.set(attributeID, in.readFloat());
				len -= 4;
			} else {
				float[] array = new float[len/4];
				for (int i=0; i<array.length; i++) {
					array[i] = in.readFloat();
					len-=4;
				}
				atts.set(attributeID, array);
			}
			break;
		case Attributes.INT:
			if (len == 4) {
				atts.set(attributeID, in.readInt());
				len -= 4;
			} else {
				int[] array = new int[len/4];
				for (int i=0; i<array.length; i++) {
					array[i] = in.readInt();
					len-=4;
				}
				atts.set(attributeID, array);
			}
			break;
		case Attributes.LONG:
			if (len == 8) {
				atts.set(attributeID, in.readLong());
				len -= 8;
			} else {
				long[] array = new long[len/8];
				for (int i=0; i<array.length; i++) {
					array[i] = in.readLong();
					len-=8;
				}
				atts.set(attributeID, array);
			}
			break;
		case Attributes.SHORT:
			if (len == 2) {
				atts.set(attributeID, in.readShort());
				len -= 2;
			} else {
				short[] array = new short[len/2];
				for (int i=0; i<array.length; i++) {
					array[i] = in.readShort();
					len-=2;
				}
				atts.set(attributeID, array);
			}
			break;
		case Attributes.STRING:
			List strings = new ArrayList();
			while(len > 0) {
				String s = in.readUTF();
				in.readByte(); // skip NULL termination
				strings.add(s);
				// len (short) + null (byte)
				len -= WBXMLParser.stringUTFLength(s) + 2 + 1;
			}
			if (strings.size() == 1) {
				atts.set(attributeID, (String)strings.get(0));
			} else {
				String[] array = new String[strings.size()];
				array = (String[])strings.toArray(array);
				atts.set(attributeID, array);
			}
			break;
		default:
			System.err.println("Unknown extension: Opaque " + len + " tagID: " + tagID
					+ " attributeID: " + attributeID);
		}
		if (len > 0) {
			System.err.println("Skipping "+len+" unused OPAQUE bytes...");
			while (len > 0) {
				in.readByte();
				len--;
			}
		}
	}

}

File: bin/benchmarks_5th/projects/freehep/native/freehep-wbxml/src/main/c++/DefaultExtensionHandler.cpp
#include <iostream>
#include <istream>

#include <WBXML/DefaultExtensionHandler.h>

using namespace std;
using namespace WBXML;

bool DefaultExtensionHandler::ext(unsigned int i, unsigned int tagID,
		int attributeID, MutableAttributes& atts, vector<wstring>& value) {
	if (debug) cout << endl << "Ext "<< i << " tagID: "<< tagID<< " attributeID: "
			<< attributeID;
	return true;
}

bool DefaultExtensionHandler::extI(unsigned int i, wstring s,
		unsigned int tagID, int attributeID, MutableAttributes& atts,
		vector<wstring>& value) {
	if (debug) wcout << endl << "ExtI "<< i << " "<< s << " tagID: "<< tagID
			<< " attributeID: "<< attributeID;
	if (i == 0) {
		atts.set(attributeID, s);
		strings.push_back(s);
	}
	return true;
}

bool DefaultExtensionHandler::extT(unsigned int i, unsigned int index,
		unsigned int tagID, int attributeID, MutableAttributes& atts,
		vector<wstring>& value) {
	if (debug) cout << endl << "ExtT "<< i << " "<< index << " tagID: "<< tagID
			<< " attributeID: "<< attributeID;
	if (i==0) {
		atts.set(attributeID, strings[index]);
	}
	return true;
}

// FIXME, no handling of arrays yet
bool DefaultExtensionHandler::opaque(unsigned int len, IReader& reader,
		unsigned int tagID, int attributeID, MutableAttributes& atts,
		vector<wstring>& value) {
	// read type
	unsigned char type;
	if (!reader.readByte(type))
		return false;
	if (debug) cout << " : Type: "<< atts.getTypeName((IAttributes::Types)type);
	len--;

	switch (type) {
	case IAttributes::BOOLEAN:
		if (len == 1) {
			bool b;
			if (!reader.readBoolean(b))
				return false;
			atts.set(attributeID, b);
			len--;
		} else {
			//			boolean[] array = new boolean[len];
			//			for (int i=0; i<array.length; i++) {
			//				array[i] = in.readByte() != 0;
			//				len--;
			//			}
			//			atts.set(attributeID, array);
		}
		break;
	case IAttributes::BYTE:
		if (len == 1) {
			unsigned char b;
			if (!reader.readByte(b))
				return false;
			atts.set(attributeID, b);
			len--;
		} else {
			//			byte[] array = new byte[len];
			//			for (int i=0; i<array.length; i++) {
			//				array[i] = in.readByte();
			//				len--;
			//			}
			//			atts.set(attributeID, array);
		}
		break;
	case IAttributes::CHAR:
		if (len == 2) {
			wchar_t c;
			if (!reader.readChar(c))
				return false;
			atts.set(attributeID, c);
			len-=2;
		} else {
			//			char[] array = new char[len/2];
			//			for (int i=0; i<array.length; i++) {
			//				array[i] = in.readChar();
			//				len-=2;
			//			}
			//			atts.set(attributeID, array);
		}
		break;
	case IAttributes::DOUBLE:
		if (len == 8) {
			double d;
			if (!reader.readDouble(d))
				return false;
			atts.set(attributeID, d);
			len -= 8;
		} else {
			//			double[] array = new double[len/8];
			//			for (int i=0; i<array.length; i++) {
			//				array[i] = in.readDouble();
			//				len-=8;
			//			}
			//			atts.set(attributeID, array);
		}
		break;
	case IAttributes::FLOAT:
		if (len == 4) {
			float f;
			if (!reader.readFloat(f))
				return false;
			atts.set(attributeID, f);
			len -= 4;
		} else {
			//			float[] array = new float[len/4];
			//			for (int i=0; i<array.length; i++) {
			//				array[i] = in.readFloat();
			//				len-=4;
			//			}
			//			atts.set(attributeID, array);
		}
		break;
	case IAttributes::INT:
		if (len == 4) {
			int i;
			if (!reader.readInt(i))
				return false;
			atts.set(attributeID, i);
			len -= 4;
		} else {
			//			int[] array = new int[len/4];
			//			for (int i=0; i<array.length; i++) {
			//				array[i] = in.readInt();
			//				len-=4;
			//			}
			//			atts.set(attributeID, array);
		}
		break;
	case IAttributes::LONG:
		if (len == 8) {
			int64 l;
			if (!reader.readLong(l))
				return false;
			atts.set(attributeID, l);
			len -= 8;
		} else {
			//			long[] array = new long[len/8];
			//			for (int i=0; i<array.length; i++) {
			//				array[i] = in.readLong();
			//				len-=8;
			//			}
			//			atts.set(attributeID, array);
		}
		break;
	case IAttributes::SHORT:
		if (len == 2) {
			short s;
			if (!reader.readShort(s))
				return false;
			atts.set(attributeID, s);
			len -= 2;
		} else {
			//			short[] array = new short[len/2];
			//			for (int i=0; i<array.length; i++) {
			//				array[i] = in.readShort();
			//				len-=2;
			//			}
			//			atts.set(attributeID, array);
		}
		break;
	case IAttributes::STRING: {
		vector<wstring> vs;
		while (len != 0) {
			wstring s;
			unsigned short utflen;
			if (!reader.readUTF(s, utflen))
				return false;
			unsigned char b;
			if (!reader.readByte(b))
				return false; // skip NULL termination
			vs.push_back(s);
			len -= utflen + 2+ 1;
		}
		if (vs.size() == 1) {
			atts.set(attributeID, vs[0]);
		} else {
			//			String[] array = new String[vs.size()];
			//			array = (String[])vs.toArray(array);
			//			atts.set(attributeID, array);
		}
	}
		break;
	default:
		cerr << endl << "ERROR No opaque handler for attributeID: "<< attributeID
				<< " part of tagID: "<< tagID << endl;
		return false;
	}
	if (debug) cout << " : Value";
	
	if (len != 0) {
		cerr << endl << "Skipping "<< len << " unused OPAQUE bytes..."<< endl;
		while (len != 0) {
			unsigned char b;
			if (!reader.readByte(b))
				return false;
			len--;
		}
	}
	return true;
}
